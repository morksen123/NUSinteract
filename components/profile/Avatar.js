import React, { useState, useContext, useEffect } from 'react';

import {StyleSheet, Image, View, Alert, Text } from 'react-native';

import { supabase } from '../../utils/supabase';

import * as ImagePicker from 'expo-image-picker';

import { UserContext } from '../../contexts/userContext';

import OutlinedButton from '../Buttons/OutlinedButton';


/* polyfills */
/** URL polyfill */
import 'react-native-url-polyfill/auto';
import { ActivityIndicator } from 'react-native-paper';

const Avatar = ({ data }) => {

  const [avatarUrl, setAvatarUrl] = useState(data[0].avatar_url);
  const [loading, setLoading] = useState(true)

  const { user } = useContext(UserContext);


  /*
  ** Allows picking of image from photo gallery
  */
  const pickImage = async () => {

    setLoading(true)
      
    // No permissions request is necessary for launching the image library
    let result = await ImagePicker.launchImageLibraryAsync({
      mediaTypes: ImagePicker.MediaTypeOptions.All,
      allowsEditing: true,
      aspect: [4, 3],
      quality: 1,
    });

    let avatarUrl = '';
    // uploads image to supabase storage if there's no error with imagePicker
    if (!result.cancelled) {

      alert('Uploading Profile Picture...')

      const ext = result.uri.substring(result.uri.lastIndexOf(".") + 1);

      const fileName = result.uri.replace(/^.*[\\\/]/, "");

      var formData = new FormData();
      formData.append("files", {
          uri: result.uri,
          name: fileName, 
          type: result.type ? `image/${ext}` : `video/${ext}`,
      })

      const { data, error } = await supabase.storage
        .from("avatars")
        .upload(fileName, formData)

      if (data) {
        setAvatarUrl(data.Key); 
        avatarUrl = data.Key
      }

      setLoading(false)
    }
    
    // updates avatarUrl associated with profile 
    const { data, error } = await supabase.from('users').upsert({
        id: user.id, 
        avatar_url: avatarUrl
    })

  };

  return (
    <View style={{ flex: 1, alignItems: 'center', justifyContent: 'center' }}>

      {
        <View>
         {loading &&

         <ActivityIndicator
              style={{
                position: 'absolute',
                left: 0,
                right: 0,
                top: 0,
                bottom: 0,
                alignItems: 'center',
                justifyContent: 'center'
              }}
          /> }

          <Image 
            source={{ uri: `https://aqeopdkkfhradtlezpil.supabase.co/storage/v1/object/public/${avatarUrl}`}} 
            style={{width: 200, height: 200, borderRadius: 200 / 2}} 
            onLoad={() => setLoading(false)}
          />
          
        </View>
          
      }

      <View style={styles.button}>
        <OutlinedButton icon="person-circle" onPress={pickImage}>Choose Profile Picture</OutlinedButton>
        <Text style={styles.username}>{user.user_metadata.username} </Text> 
      </View>

    </View>
  );
}

export default Avatar;


const styles = StyleSheet.create({

  button:{
    marginTop: 10
  },

  username : {
    fontSize: 30, 
    textAlign: 'center',
    marginTop: 15,
    fontWeight: '600'
  },

});