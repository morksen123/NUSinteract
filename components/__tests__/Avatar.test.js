import React from "react";
import renderer from 'react-test-renderer';
import Avatar from "../profile/Avatar";


test('renders correctly', () => {
    const data = [{ name: 'dummy', avatar_url: 'avatars/blank-profile-picture-973460_1280.webp'}]
    const tree = renderer.create(<Avatar data={data}/>).toJSON();
    expect(tree).toMatchSnapshot();
  });