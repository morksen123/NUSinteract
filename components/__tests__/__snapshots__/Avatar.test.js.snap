// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View>
    <View
      accessibilityRole="progressbar"
      accessibilityState={
        Object {
          "busy": true,
        }
      }
      accessible={true}
      style={
        Array [
          Object {
            "alignItems": "center",
            "justifyContent": "center",
          },
          Object {
            "alignItems": "center",
            "bottom": 0,
            "justifyContent": "center",
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
        ]
      }
    >
      <View
        collapsable={false}
        style={
          Object {
            "height": 24,
            "opacity": 1,
            "width": 24,
          }
        }
      >
        <View
          collapsable={false}
          style={
            Object {
              "alignItems": "center",
              "bottom": 0,
              "justifyContent": "center",
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <View
            collapsable={false}
            style={
              Object {
                "height": 24,
                "transform": Array [
                  Object {
                    "rotate": "45deg",
                  },
                ],
                "width": 24,
              }
            }
          >
            <View
              collapsable={false}
              style={
                Object {
                  "height": 12,
                  "overflow": "hidden",
                  "width": 24,
                }
              }
            >
              <View
                collapsable={false}
                style={
                  Object {
                    "height": 24,
                    "transform": Array [
                      Object {
                        "translateY": 0,
                      },
                      Object {
                        "rotate": "-165deg",
                      },
                    ],
                    "width": 24,
                  }
                }
              >
                <View
                  collapsable={false}
                  style={
                    Object {
                      "height": 12,
                      "overflow": "hidden",
                      "width": 24,
                    }
                  }
                >
                  <View
                    collapsable={false}
                    style={
                      Object {
                        "borderColor": "#6200ee",
                        "borderRadius": 12,
                        "borderWidth": 2.4,
                        "height": 24,
                        "width": 24,
                      }
                    }
                  />
                </View>
              </View>
            </View>
          </View>
        </View>
        <View
          collapsable={false}
          style={
            Object {
              "alignItems": "center",
              "bottom": 0,
              "justifyContent": "center",
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <View
            collapsable={false}
            style={
              Object {
                "height": 24,
                "transform": Array [
                  Object {
                    "rotate": "45deg",
                  },
                ],
                "width": 24,
              }
            }
          >
            <View
              collapsable={false}
              style={
                Object {
                  "height": 12,
                  "overflow": "hidden",
                  "top": 12,
                  "width": 24,
                }
              }
            >
              <View
                collapsable={false}
                style={
                  Object {
                    "height": 24,
                    "transform": Array [
                      Object {
                        "translateY": -12,
                      },
                      Object {
                        "rotate": "345deg",
                      },
                    ],
                    "width": 24,
                  }
                }
              >
                <View
                  collapsable={false}
                  style={
                    Object {
                      "height": 12,
                      "overflow": "hidden",
                      "width": 24,
                    }
                  }
                >
                  <View
                    collapsable={false}
                    style={
                      Object {
                        "borderColor": "#6200ee",
                        "borderRadius": 12,
                        "borderWidth": 2.4,
                        "height": 24,
                        "width": 24,
                      }
                    }
                  />
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
    <Image
      onLoad={[Function]}
      source={
        Object {
          "uri": "https://aqeopdkkfhradtlezpil.supabase.co/storage/v1/object/public/avatars/blank-profile-picture-973460_1280.webp",
        }
      }
      style={
        Object {
          "borderRadius": 100,
          "height": 200,
          "width": 200,
        }
      }
    />
  </View>
  <View
    style={
      Object {
        "marginTop": 10,
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#9E89FE",
            "borderRadius": 20,
            "borderWidth": 2,
            "flexDirection": "row",
            "justifyContent": "center",
            "margin": 4,
            "paddingHorizontal": 12,
            "paddingVertical": 6,
          },
          false,
        ]
      }
    >
      <Text />
      <Text
        style={
          Object {
            "color": "white",
          }
        }
      >
        Choose Profile Picture
      </Text>
    </View>
  </View>
</View>
`;
